chain/client/src/sync/sync_actor.rs

```diff
@@ -107,3 +148,40 @@ impl actix::Handler<WithSpanContext<StateSyncResponse>> for SyncActor {
         self.handle_network_sync_message(msg);
     }
 }
+
+impl actix::Handler<WithSpanContext<ApplyStatePartsRequest>> for SyncActor {
+    type Result = ();
+    #[perf]
+    fn handle(
+        &mut self,
+        msg: WithSpanContext<ApplyStatePartsRequest>,
+        _ctx: &mut Self::Context,
+    ) -> Self::Result {
+        let (_span, msg) = handler_debug_span!(target: "sync", msg);
+        let shard_id = msg.shard_uid.shard_id as ShardId;
+        match self.clear_flat_state(&msg) {
+            Err(err) => {
+                self.senders.client_adapter.send(ApplyStatePartsResponse {
+                    apply_result: Err(err),
+                    shard_id,
+                    sync_hash: msg.sync_hash,
+                });
+                return;
+            }
+            Ok(false) => {
+                // Can't panic here, because that breaks many KvRuntime tests.
+                // TODO: figure out how to fix things so that we can panic here
+                tracing::error!(target: "client", shard_uid = ?msg.shard_uid, "Failed to delete Flat State, but proceeding with applying state parts.");
+            }
+            Ok(true) => {
+                tracing::debug!(target: "client", shard_uid = ?msg.shard_uid, "Deleted all Flat State");
+            }
+        }
+        let result = self.apply_parts(&msg);
+        self.senders.client_adapter.send(ApplyStatePartsResponse {
+            apply_result: result,
+            shard_id,
+            sync_hash: msg.sync_hash,
+        });
+    }
+}
```

Comment:
How do you feel about implementing the new logic in the state sync actors without connecting it to the old logic? 
The new sync logic does not report back to `Client` every time it applies a part.

We can keep the old logic running as it is and switch to the new one when we have it functional.
We will delete the `SyncJobActor` thread and the whole logic in client will be simplified to wait for the `SyncDone` message from `SyncActor`

Reply:
Yeah that sounds like a good idea. In the meantime I think this PR kind of sets up more of the scaffolding that we can use for that, and keeps most of the old sync logic in the client actor. So since the only logic we're moving into the `SyncActor` in this PR is the `SyncJobsActor` logic on applying parts, we can just remove that pretty easily when it's no longer needed, and in the meantime with this PR we at least get parallelized state part application when you download state parts from S3